### PR TITLE
Create `crypto` module

### PR DESCRIPTION
**Pull request by tcharding**
Done as part of [util flattening](https://github.com/rust-bitcoin/rust-bitcoin/issues/639).

Create a `crypto` module and move into it (out of `util`):
- ecdsa
- schnorr
- key

After review, this PR now includes some type re-names
- EcdsaSig -> ecdsa::Signature
- SchnorrSig -> schnorr::Signature
- EcdsaSigError -> ecdsa::Error
- SchnorrSigError -> schnorr::Error
- InvalidSchnorrSigSize -> InvalidSignatureSize  (this is an error enum variant)

### CODE REVIEW

**Comment by apoelstra** [comment]
Musing whether this should be its own crate ... oftentimes I'm working on stuff where everything but the crypto is "easy" and it'd be nice to minimize my dep tree.

**Comment by Kixunil** [comment]
I wonder if this has the same problem as `bips`. Doesn't feel like it but my intuition may be failing me. Will think about it more.

**Comment by apoelstra** [comment]
I think `crypto` is a sensible category and has useful connotations (e.g. "you probably don't need to use things in this module" "here be dragons").

**Comment by tcharding** [comment]
> Musing whether this should be its own crate ... oftentimes I'm working on stuff where everything but the crypto is "easy" and it'd be nice to minimize my dep tree.

Is this PR still a step in the right direction, first move everything into a module and then later move it into a separate crate?

**Review by apoelstra — APPROVED** [review]
ACK 607ec4cfa40da7d80278cc72fae09ab1dcd7923e

**Comment by tcharding** [comment]
Should `sighash` be in `crypto` as well?

**Comment by apoelstra** [comment]
Can we make this also move `key` into `crypto`?

**Comment by tcharding** [comment]
> Can we make this also move `key` into `crypto`?

Lol, I was busy writing https://github.com/rust-bitcoin/rust-bitcoin/pull/1265#issuecomment-1246013132 while you posted.

**Comment by tcharding** [comment]
Now includes `key`!

**Review by apoelstra — DISMISSED** [review]
ACK 2bfdb490b268c8ea5101f35be5d84158bd3cdcfc

File: bitcoin/src/lib.rs [hunk 1]
```diff
@@ -104,13 +105,23 @@ pub use crate::blockdata::transaction::{self, OutPoint, Sequence, Transaction, T
 pub use crate::blockdata::witness::{self, Witness};
 pub use crate::blockdata::{constants, opcodes};
 pub use crate::consensus::encode::VarInt;
+pub use crate::crypto::ecdsa::{self, EcdsaSig, EcdsaSigError};
+pub use crate::crypto::hash::{self, bitcoin_merkle_root, bitcoin_merkle_root_inline};
+pub use crate::crypto::key::{self, KeyPair, PrivateKey, PublicKey, XOnlyPublicKey};
+pub use crate::crypto::merkleblock::{self, MerkleBlock, MerkleBlockError, PartialMerkleTree};
```
Comment (Kixunil) [inline bitcoin/src/lib.rs:111]: Should we actually re-export whole modules, not just types?

File: bitcoin/src/lib.rs [hunk 2]
```diff
@@ -104,13 +105,23 @@ pub use crate::blockdata::transaction::{self, OutPoint, Sequence, Transaction, T
 pub use crate::blockdata::witness::{self, Witness};
 pub use crate::blockdata::{constants, opcodes};
 pub use crate::consensus::encode::VarInt;
+pub use crate::crypto::ecdsa::{self, EcdsaSig, EcdsaSigError};
+pub use crate::crypto::hash::{self, bitcoin_merkle_root, bitcoin_merkle_root_inline};
```
Comment (Kixunil) [inline bitcoin/src/lib.rs:109]: Reexporting this seems too much. They are quite niche functions.

File: bitcoin/src/lib.rs [hunk 3]
```diff
@@ -104,13 +105,23 @@ pub use crate::blockdata::transaction::{self, OutPoint, Sequence, Transaction, T
...
+pub use crate::crypto::schnorr::{
+    self, SchnorrSig, SchnorrSigError, TapTweak, TweakedKeyPair, TweakedPublicKey,
+    UntweakedKeyPair, UntweakedPublicKey,
+};
+pub use crate::crypto::taproot::{
+    self, ControlBlock, FutureLeafVersion, LeafVersion, NodeInfo, ScriptLeaf, TaprootBuilder,
+    TaprootBuilderError, TaprootError, TaprootMerkleBranch, TaprootSpendInfo, TAPROOT_ANNEX_PREFIX,
+    TAPROOT_CONTROL_BASE_SIZE, TAPROOT_CONTROL_MAX_NODE_COUNT, TAPROOT_CONTROL_MAX_SIZE,
+    TAPROOT_CONTROL_NODE_SIZE, TAPROOT_LEAF_MASK, TAPROOT_LEAF_TAPSCRIPT,
+};
```
Comment (Kixunil) [inline bitcoin/src/lib.rs:121]: It seems too much to reexport these.

**Review by Kixunil — COMMENTED** [review]
Mostly OK but I think this reexports too many items. AFAIK the only items used very often are keys and they should be reexported, others probably not.

**Comment by apoelstra** [comment]
On second thought, I agree with @Kixunil. Probably the only things on this list we should obviously export are the key types. I'd also like to reexport `EcdsaSig`.

**Comment by tcharding** [comment]
I re-exported everything with the idea to de-couple the API from the directory structure. I was basing the idea on this comment https://github.com/rust-bitcoin/rust-bitcoin/issues/639#issuecomment-895625067 from @thomaseizinger. I don't currently work on any code that consumes this library but when I did I found it annoying to have to remember where things were, especially something like `TAPROOT_LEAF_MASK` - that is already obviously unique across the `rust-bitcoin` crate. Open to learning on this issue but I'd like to get consensus so I know what we are aiming for since this effects all the modules.

**Comment by thomaseizinger** [comment]
Personally, what I usually try to balance is:

- Minimize number of imports
- Minimize verbosity in the code
- Minimize ambiguity in the code

One can have low number of imports if you fully qualify everything but that is verbose yet unambiguous.

One can have low verbosity by importing everything but that creates ambiguity and high number of imports. For example `fmt::Result` is clearer than just `Result`.

A good example is to import `std::fmt` when implementing `std::fmt::Display`:

- Only one import
- `impl fmt::Display for Foo` is not too verbose over `impl Display for Foo`
- `fmt::Result` is clearer over `Result` and doesn't clash with other types

This is the consumer side. From a crate author's perspective, you now need to look at, what types are often used together. What are they called? Are they unambigous (across the Rust ecosystem really) so they can be referenced mostly by just their name?

Many types are often unambigous when qualified by their crate name, i.e. `bitcoin::Amount`. Submodules essentially allow you further shape your crate's API by trading away a bit of ambiguity (there are likely multiple `crypto` modules across the ecosystem) for minimizing imports by saying things like `crypto::Key` or `crypto::Signature`.

One more thing: crypto seems like a good module to group a bunch of code together but I think it is not a great export candidate. Perhaps a better idea would be to make it private and just reexport its modules underneath / some types based on the above thinking.

Hope this helps clarify what I meant in the above referenced comment 😊

**Comment by apoelstra** [comment]
My rule of thumb is that if you can sensibly talk about e.g. a "bitcoin public key" in the vernacular then we should have a `bitcoin::PublicKey`. So the key types, address types, blocks, transactions, etc., should be re-exported. `Amount` also meets this criteria, as do `EcdsaSignature` and `SchnorrSignature` though those are iffier.

Something like `TAPROOT_LEAF_MASK` seems like a pretty obscure constant which would just create noise in the documentation. I wouldn't mind having a `bitcoin::constants` module where all these things were re-exported.

Then of course there's a lot of in-between stuff which are judgement calls really.

**Comment by tcharding** [comment]
Cool, thanks for both the explanations - I'll try and internalise these and re-work the exports in this PR.

**Comment by tcharding** [comment]
Trimmed the exports down to 
```rust
pub use crate::crypto::{hash, merkleblock, schnorr, taproot};
pub use crate::crypto::ecdsa::{self, EcdsaSig, EcdsaSigError};
pub use crate::crypto::key::{self, PrivateKey, PublicKey};
```

**Comment by tcharding** [comment]
Process fail, somehow I didn't commit or push the changes when I wrote that message above :(

**Comment by tcharding** [comment]
Fixed rustdoc tests.

**Comment by Kixunil** [comment]
Isn't `EcdsaSigError` mainly used by its methods, thus not needing reexporting? Also I like `ecdsa::Signature` more.

**Comment by apoelstra** [comment]
Since you have to rebase anyway to deal with CI, I agree that we should rename `EcdsaSig` and `EcdsaSigError` to drop the prefix, and make the user type `ecdsa::` instead.

**Comment by tcharding** [comment]
> Fixed rustdoc tests.

Holy S*** I was hopeless yesterday, I didn't commit the rustdoc changes.

**Comment by tcharding** [comment]
> Isn't EcdsaSigError mainly used by its methods, thus not needing reexporting?

Its used in `psbt::serialize`

**Review by apoelstra — DISMISSED** [review]
ACK 9c4277b7ce1084dd48be624df815c222e9c675ef

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Review by apoelstra — DISMISSED** [review]
ACK 9dfd269088b063815aed4c94990809461c2f0e45

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Comment by tcharding** [comment]
Total fail on this PR (git index issue I mentioned elsewhere). I've fixed it up but the PR now does not include formmatting changes, sorry I can't do rustfmt on holidays :)

My apologies for adding review burden by my mistake.

**Review by apoelstra — APPROVED** [review]
ACK 10f2e957fd27a969e9fb1325988c931c5232abd7

**Comment by tcharding** [comment]
I had an unrelated change in this PR removing the re-export of `sighash` from crate root, I've removed that change while rebasing.

**Review by apoelstra — DISMISSED** [review]
ACK 6c26ad6ac3fc888cb811ab4c1df332d7d77ed400

**Review by sanket1729 — APPROVED** [review]
ACK 6c26ad6ac3fc888cb811ab4c1df332d7d77ed400. `taproot` is not really crypto. 

It would make sense to split the methods that deal with crypto into the crypto module. It can either be its own module or perhaps can be justified in blockdata. 

I am not really sure, there is some crypto here, but most of it is data structures. I ACKed because this is a strict improvement over the current master.

**Comment by tcharding** [comment]
Cheers @sanket1729, I'll add an item to my todo list to look into your comment, could be done as part of making `crypto` a separate crate, which seems to be on the cards.

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Review by apoelstra — APPROVED** [review]
ACK 38e6dee4e71f869a7c4f4a13677b5d9246ba8b06

**Comment by tcharding** [comment]
Changes in force push:
- rebase
- reduced the diff by minimising changes in import statements (a few files that are not currently formatted have messy imports, I reduced the changes to imports in these files to make review easier)

**Comment by apoelstra** [comment]
range-diff is still fine. There is a lot of deleted diff which is easy to read.

**Review by apoelstra — DISMISSED** [review]
ACK 698176da0076a0b6b9324d14ba5cfd71ae55ec1c

**Comment by tcharding** [comment]
Rebased to fix trivial merge conflict in import statements.

**Review by apoelstra — APPROVED** [review]
ACK bc0452297692d2be87f7c3edf9b4a52dddcbdab6

**Comment by tcharding** [comment]
Rebased!

**Review by apoelstra — APPROVED** [review]
ACK 45623fb9600b77a0f4c11d2136f8f056da981d19

**Comment by tcharding** [comment]
FTR while working on this one I've done #1373 and #1374, which I believe are superior to the module moves first presented in this PR (`taproot` and `merkleblock`).

**Comment by tcharding** [comment]
Changes in force push are non-trivial:
- rebased on master
- `util::hash` module no-longer exits on master (was re-named to `src/merkle_tree.rs`)
- `taproot` module is no longer moved in this PR (handled in #1373) 
- `merkleblock` module is no longer moved in this PR (handled in #1374)
- re-wrote PR description

**Review by sanket1729 — DISMISSED** [review]
ACK 4252ab59c70f620bff521e135431ae159689e116. @tcharding, please continue to aggressively (ab)use the P-high label. Really helps in prioritizing reviews.

**Comment by apoelstra** [comment]
Needs rebase again. (Or at least, github thinks so.)

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Comment by tcharding** [comment]
And this one too please @sanket1729, we are so close ...

**Review by apoelstra — DISMISSED** [review]
ACK 132d24bbd64924014a6009b8b55b502d9bd08a85

**Comment by tcharding** [comment]
Rebase, no other changes.

**Review by apoelstra — APPROVED** [review]
ACK 69b397a006c9b1260a3b4023098f3b3abd1ce8c7

**Review by apoelstra — APPROVED** [review]
ACK 2df51dae159b1e1fca854fb147377a489b9c5d72

**Comment by tcharding** [comment]
Can I get another ack please @sanket1729, to save you copying the last commit hash you ack'ed here it is `git range-diff 4252ab5...@`

**Review by sanket1729 — APPROVED** [review]
ACK 2df51dae159b1e1fca854fb147377a489b9c5d72

**Comment by tcharding** [comment]
BOOM! Thanks bro

**Pull request merged**